core/common/src/main/java/alluxio/wire/MetricValue.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.wire;
+
+import com.google.common.base.Objects;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Represents the value of a metrics property, only one of longValue or doubleValue should be set.
```

Comment:
it would be nice to make this immutable so that we can enforce the invariant that only one of the fields is set. To make it work with json/jackson, the construction can be like

```java
@JsonCreator
private MetricValue(@JsonProperty("doubleValue") doubleValue, @JsonProperty("longValue") longValue) {
  Preconditions.checkState(doubleValue == null || longValue == null, "only one of longValue and doubleValue can be set");
  mDoubleValue = doubleValue;
  mLongValue = longValue;
}

public MetricValue forLong(long longValue) {
  return new MetricValue(null, longValue);
}

public MetricValue forDouble(double doubleValue) {
  return new MetricValue(doubleValue, null);
}
```

Reply:
Operator "==" could not be applied to double and null, long and null.
It seems that we cannot control whether user set one field or more.

```  java
@JsonCreator
  private MetricValue(
      @JsonProperty("doubleValue") double doubleValue,
      @JsonProperty("longValue") long longValue){
    Preconditions.checkState(doubleValue == null || longValue == null, "only one of longValue and doubleValue can be set");
    // Operator "==" could not be applied to double and null, long and null.
    mDoubleValue = doubleValue;
    mLongValue = longValue;
  }

  public MetricValue forLong(long longValue) {
    return new MetricValue(null, longValue); //null value cannot be applied to double.
  }

  public MetricValue forDouble(double doubleValue) {
    return new MetricValue(doubleValue, null);//null value cannot be applied to long
  }
```

Reply:
can we use `Double` and `Long` instead?

Reply:
Yes! Double and Long work! Thanks for the info!
I still have one question about the why we use Jackson here.
I am not quite understand why me need to use Jackson to create Object from Json by using @JsonCreator and @JsonProperty. Without these two annotations, the Metric Value constructor can still work and can still use Preconditions.checkState to make sure only one of Double and Long has been set.